src/emulator/storage/upload.ts

```diff
@@ -0,0 +1,188 @@
+export type Upload = {
+  id: string;
+  bucketId: string;
+  objectId: string;
+  type: UploadType;
+  // Path to where the file is stored on disk. May contain incomplete data if
+  path: string;
+  status: UploadStatus;
+  metadata: IncomingMetadata;
+  size: number;
+  authorization?: string;
+};
+
+export enum UploadType {
+  MULTIPART,
+  RESUMABLE,
+}
+
+/** The status of an upload. Multipart uploads can only ever be FINISHED. */
+export enum UploadStatus {
+  ACTIVE,
+  CANCELLED,
+  FINISHED,
+}
+
+/** Request object for {@link UploadService#multipartUpload}. */
+export type MultipartUploadRequest = {
+  bucketId: string;
+  objectId: string;
+  metadataRaw: string;
+  dataRaw: string;
+  authorization?: string;
+};
+
+/** Request object for {@link UploadService#startResumableUpload}. */
+export type StartResumableUploadRequest = {
+  bucketId: string;
+  objectId: string;
+  metadataRaw: string;
+  contentType: string;
+  authorization?: string;
+};
+
+/** Error that signals a resumable upload that's expected to be active is not. */
+export class UploadNotActiveError extends Error {}
+
+/** Error that signals a resumable upload is not cancellable.  */
+export class NotCancellableError extends Error {}
+
+/**
+ * Service that handles byte transfer and maintains state for file uploads.
+ *
+ * New file uploads will be persisted to a temp staging directory which will not
+ * survive across emulator restarts. Clients are expected to move staged files
+ * to a more permanent location.
+ */
+export class UploadService {
+  private _uploads!: Map<string, Upload>;
+  constructor(private _persistence: Persistence) {
+    this.reset();
+  }
+
+  /** Resets the state of the UploadService. */
+  public reset(): void {
+    this._uploads = new Map();
+  }
+
+  /**
+   * Handles a multipart file upload which is expected to have the entirety of
+   * the file's contents in a single request.
+   */
+  public multipartUpload(request: MultipartUploadRequest): Upload {
+    const data = Buffer.from(request.dataRaw);
+    const upload = this.initMultipartUpload(request, data.byteLength);
+    this._persistence.deleteFile(upload.path, /* failSilently = */ true);
+    this._persistence.appendBytes(upload.path, data);
+    return upload;
+  }
+
+  private initMultipartUpload(request: MultipartUploadRequest, sizeInBytes: number): Upload {
+    const id = uuidV4();
+    const upload: Upload = {
+      id: uuidV4(),
+      bucketId: request.bucketId,
+      objectId: request.objectId,
+      type: UploadType.MULTIPART,
+      path: this.stagingFileName(id, request.bucketId, request.objectId),
+      status: UploadStatus.FINISHED,
+      metadata: JSON.parse(request.metadataRaw),
+      size: sizeInBytes,
+      authorization: request.authorization,
+    };
+    this._uploads.set(upload.id, upload);
+
+    return upload;
+  }
+
+  /**
+   * Initializes a new ResumableUpload.
+   */
+  public startResumableUpload(request: StartResumableUploadRequest): Upload {
```

Comment:
Naming nit: for consistency, either both `init` or both `start` for multipart and resumable

Reply:
Done